numpy/core/src/multiarray/arrayfunction_override.c

```diff
@@ -0,0 +1,213 @@
+                              int *num_overloaded_args_out)
+{
+    static PyObject *ndarray = NULL;
+    static PyObject *ndarray_array_function = NULL;
+
+    PyObject *iterator;
+    PyObject *arg;
+    PyObject *next_arg;
+    PyObject *saved;
+    PyObject *arg_type;
+    int num_overloaded_types = 0;
+    int num_overloaded_args = 0;
+    int j;
+    int new_class;
+    int arg_index;
+    int offset;
+
+    /* on first entry, import and cache ndarray and its __array_function__ */
+    if (ndarray == NULL) {
+        npy_cache_import("numpy.core.multiarray", "ndarray", &ndarray);
+        ndarray_array_function = PyObject_GetAttrString(ndarray,
+                                                        "__array_function__");
+    }
+
+    iterator = PyObject_GetIter(relevant_args);
+    if (iterator == NULL) {
+        return -1;
+    }
+
+    while ((arg = PyIter_Next(iterator))) {
+        new_class = 1;
+        for (j = 0; j < num_overloaded_types; j++) {
+            new_class = ((PyObject *)Py_TYPE(arg) != overloaded_types[j]);
+            if (!new_class) {
+                break;
+            }
+        }
+        if (new_class &&
+                PyArray_LookupSpecial(arg, "__array_function__") != NULL) {
+            /* save the type */
+            arg_type = (PyObject *)Py_TYPE(arg);
+            overloaded_types[num_overloaded_types] = arg_type;
+            ++num_overloaded_types;
+            Py_INCREF(arg_type);
+            /* TODO: can we safely avoid incrementing the reference counter
+             * for arg_type until we've determined we need an override? */
+            /* TODO: handle num_overloaded_types>NPY_MAXARGS */
+
+            /* insert the argument in the right position */
+            arg_index = num_overloaded_args;
+            for (j = 0; j < num_overloaded_args; j++) {
+                if (PyObject_IsInstance(
+                        arg, (PyObject *)Py_TYPE(overloaded_args[j]))) {
+                    arg_index = j;
+                    break;
+                }
+            }
+            next_arg = arg;
+            for (j = arg_index; j < num_overloaded_args + 1; j++) {
+                saved = overloaded_args[j];
+                overloaded_args[j] = next_arg;
+                next_arg = saved;
+            }
+            Py_INCREF(arg);
+            ++num_overloaded_args;
+        }
+        Py_DECREF(arg);
+    }
+    Py_DECREF(iterator);
+
+    if (PyErr_Occurred()) {
+        return -1;
+    }
+
+    /* remove argument with __array_function__ methods matching
+     * ndarray.__array_function__ */
+    offset = 0;
+    for (j = 0; j < num_overloaded_args; j++) {
+        arg = overloaded_args[j];
+        if ((PyObject *)Py_TYPE(arg) != ndarray ||
+                PyArray_LookupSpecial(arg, "__array_function__") !=
+                ndarray_array_function) {
+            overloaded_args[j - offset] = arg;
+        }
+        else {
+            ++offset;
+            Py_DECREF(arg);
```

Comment:
This logic did not catch subclasses of `ndarray` with no `__array_function__` override, and caused infinite recursion. While the diff below seems to fix it (`!=` is so much more confusing than `==` you should have used `&&` not `||`) , perhaps it would be prudent to also [guard against recursion](https://github.com/numpy/numpy/blob/v1.15.4/numpy/core/arrayprint.py#L440)

<details>

``` diff
     for (j = 0; j < num_overloaded_args; j++) {
         arg = overloaded_args[j];
-        if ((PyObject *)Py_TYPE(arg) != ndarray ||
-                PyArray_LookupSpecial(arg, "__array_function__") !=
+        if ((PyObject *)Py_TYPE(arg) == ndarray ||
+                PyArray_LookupSpecial(arg, "__array_function__") ==
                 ndarray_array_function) {
-            overloaded_args[j - offset] = arg;
-        }
-        else {
             ++offset;
             Py_DECREF(arg);
         }
+        else {
+            overloaded_args[j - offset] = arg;
+        }

```

Reply:
thanks -- good catch!